### PR TITLE
make domain optional

### LINKED ISSUE #1707: [BUG] - local deploy cannot guaranteed be done without having access to a domain
### Describe the bug

Related to #1703. TL;DR: our guide tells users to fix `/etc/hosts`, but the deployed cluster doesn't see this change. There are three ways to achieve that:

1. Have access to a domain and create a DNS entry for `--domain=...` to point to `172.18.1.100`. This is what we do for `github-actions.nebari.dev` to get our CI going:
   https://github.com/nebari-dev/nebari/blob/daecbcf37497b1b8b4e3f0f52667ababf3c023ad/.github/workflows/kubernetes_test.yaml#L95
2. Use a service like [nip.io](https://nip.io/) and set something like `--domain=172.18.1.100.nip.io` during `nebari init`. With this users no longer need to have a domain available, but still need to have access to the internet. Plus, we add an external dependency, since our guide then depends on nip.io being available.
3. Instead of using an URL, users could also use the IP directly, i.e. `--domain=172.18.1.100`. It is not as pretty as having an URL, but probably good enough for local deploy. The major advantage is that this can be done even without internet access making it the most reliable and easiest to use option of the three. Doing this, the user also doesn't need to fix `/etc/hosts`.

Regardless of what we decide we want to use, 2. and 3. (as well as our current docs) make an assumption that doesn't hold in 100% of the cases: the cluster IP is fixed at `172.18.1.100`. 

However, we don't know that. It should work most of the time, but I know for example that it doesn't work for @costrouc. Maybe he can fill in some details on when this won't be true?

Internally, we only know the correct IP after stage 4:

https://github.com/nebari-dev/nebari/blob/daecbcf37497b1b8b4e3f0f52667ababf3c023ad/nebari/deploy.py#L102-L104

Meaning, setting it upfront during `nebari init --domain` is not guaranteed to work.

### Expected behavior

I think we should allow setting no domain upfront and instead of asking the user to set up their DNS (when `--disallow-prompt` is not set)

https://github.com/nebari-dev/nebari/blob/daecbcf37497b1b8b4e3f0f52667ababf3c023ad/nebari/deploy.py#L128-L131

just use the IP we get from the load balancer.

### OS and architecture in which you are running Nebari

Linux

### How to Reproduce the problem?

Follow the [local deployment how-to](https://www.nebari.dev/docs/how-tos/nebari-local) in an environment where the load balancer doesn't use `172.18.1.100` as IP.

### Command output

_No response_

### Versions and dependencies used.

_No response_

### Compute environment

None

### Integrations

_No response_

### Anything else?

_No response_

### PR DESCRIPTION
**Pull request by pmeier**
## Reference Issues or PRs

Closes #1707.

## What does this implement/fix?

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds a feature)
- [ ] Breaking change (fix or feature that would cause existing features not to work as expected)
- [ ] Documentation Update
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build related changes
- [ ] Other (please describe):

## Testing

- [x] Did you test the pull request locally?
- [ ] Did you add new tests?

## Any other comments?

`nebari init` (with or without `--guided-init`) and `nebari deploy` work fine. However we get problems for everything else since we never write back the IP of the load balancer that we find during deploy. I think we have two options here:

1. Write back the IP back. This probably a bad idea, since a deploy would potentially change the config. Since this should only happen for local deployments, it isn't too bad though.
2. Make an empty `domain` field a sentinel. During the verification stage, we could inspect the cluster, if available, and extract the domain from there.

### CODE REVIEW

File: nebari/cli/init.py [hunk 1]
```diff
@@ -368,15 +368,20 @@ def guided_init_wizard(ctx: typer.Context, guided_init: str):
         # DOMAIN NAME
         rich.print(
             (
-                "\n\n 🪴  Great! Now you need to provide a valid domain name (i.e. the URL) to access your Nebri instance. "
-                "This should be a domain that you own.\n\n"
+                "\n\n 🪴  Great! Now you can provide a valid domain name (i.e. the URL) to access your Nebari instance. "
+                "This should be a domain that you own. FIXME Default is the IP of the load balancer\n\n"
```
Comment (pmeier) [inline nebari/cli/init.py:372]: Need to put something else here.

File: nebari/cli/init.py [hunk 2]
```diff
@@ -368,15 +368,20 @@ def guided_init_wizard(ctx: typer.Context, guided_init: str):
...
+        inputs.domain_name = (
+            questionary.text(
+                "What domain name would you like to use?",
+                qmark=qmark,
+            ).unsafe_ask()
+            or None
+        )
+
+        print("#" * 60)
+        print(repr(inputs.domain_name))
```
Comment (pmeier) [inline nebari/cli/init.py:384]: Oops

```suggestion
```

File: nebari/cli/init.py [hunk 3]
```diff
@@ -545,6 +551,8 @@ def if_used(key, model=inputs, ignore_list=["cloud_provider"]):
                 if isinstance(value, bool) and value:
                     return b.format(key=key, value=value).replace("_", "-")
 
+        print(inputs.dict())
+
```
Comment (pmeier) [inline nebari/cli/init.py:555]: Oops

```suggestion
```

**Comment by iameskild** [comment]
Hey @pmeier, what is the status of this PR?

**Comment by viniciusdc** [comment]
Hi @pmeier, regarding what we discussed during the last community meeting, for the destroy part, if you would like the load balancer address (or any information from the resources) to be exposed, right now, we are getting it from terraform [here](https://github.com/nebari-dev/nebari/blob/develop/src/_nebari/template/stages/04-kubernetes-ingress/outputs.tf) we then use [this](https://github.com/nebari-dev/nebari/blob/a6d032f532e39bbd360955f29f6a0b5cc8c87b11/src/_nebari/provider/terraform.py#L157-L166) function to catch all the outputs (that terraform saves under a json file)

**Comment by viniciusdc** [comment]
Also, it might be interesting to run the Kubernetes CI on this PR as well, so that we can see the error message as well. I am a bit curious to see where terraform might be complaining about the address, as this should only be failing in the schema validation

**Comment by costrouc** [comment]
Closing since implemented in https://github.com/nebari-dev/nebari/pull/1833

**Pull request closed without merging**